### PR TITLE
KAFKA-3073: Add topic regex support for Connect sinks

### PR DESCRIPTION
**Pull request by jklukas**
There are more methods that had to be touched than I anticipated when writing [the KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-215%3A+Add+topic+regex+support+for+Connect+sinks).

The implementation here is now complete and includes a test that verifies that there's a call to `consumer.subscribe(Pattern, RebalanceHandler)` when `topics.regex` is provided.

### CODE REVIEW

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (jklukas) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java]: This sentence about Pattern is pulled verbatim from the docs for RegexRouter for consistency.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (jklukas) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: In order not to change the existing public signature, I've added an additional parameter for the `topicsRegex` and maintained a form without that parameter that passes in null.

The callers of this function all start with some subclass of `ConnectorConfig`, then pull out maxTasks and sinkTopics from it. It might be more elegant to have the signature here be simply `connectorTaskConfigs(String connName, ConnectorConfig config)` and then pull out the necessary contents here rather than making that the caller's responsibility. Any reason not to do that?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (jklukas) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: This signature is no longer invoked anywhere within this project, but I keep it here since it's a public method that other projects could conceivably be invoking.

**Comment by jklukas** [comment]
Is there a way to force a retest? The error `Could not determine java version from '9.0.1'.` sounds like it's related to the Jenkins environment rather than any changes here. All tests were passing before the most recent commit, which only modified html.

**Comment by jklukas** [comment]
This should be ready for review now, though let me know if the expectation is to hold off on this until we have agreement and approval on [the KIP](https://cwiki.apache.org/confluence/display/KAFKA/KIP-215%3A+Add+topic+regex+support+for+Connect+sinks).

cc @ewencp @ijuma

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (ewencp) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: It's fine to remove it. Nothing in `connect-runtime` is public, supported API. Only `connect-api` is public and it is expected that changes to APIs marked as public in Java will be made since you need to mark them public to use them from other packages.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java
Comment (ewencp) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java]: Seems like a reasonable refactoring. It likely ended up this way just due to the way this was originally developed from the prototype and with just 2 things being pulled out of the `ConnecterConfig` refactoring didn't really matter. The main drawback I can think of is possibly re: testing where tests using this code when mocked might be simpler with more parameters as the full ConnectorConfig would not be needed (e.g. see `DistributedHerderTest`). I haven't tried refactoring though, so I'm not sure how bad that would be or what would be lost if we didn't fully validate the values being used if we just allowed any `ConnectorConfig`.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: This could be `ConfigException` instead. With the `ConnectException` hierarchy we get a little bit stuck because we can't have `ConnectException` be the base of all our exceptions and not redefine some. There are some from core Kafka that are fine if triggered from within the framework, and `ConfigException` falls under that category -- the biggest value aside from standardization that asking for `ConnectException`/`RetriableException` from connectors/tasks is that it is a signal for how to handle the exception, but the runtime code doesn't require that everything inherits from those (and in fact we handle other exceptions from connectors as well, we just would prefer that the connector think about how it wants the framework to handle possible errors).

**Review by ewencp — COMMENTED** [review]
A couple of minor comments but mostly looks to be in good shape

**Comment by jklukas** [comment]
Went ahead with the refactor to `connectorTaskConfigs(String connName, ConnectorConfig config)` and got tests passing again.

I also switched to ConfigException as @ewencp suggested.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Would it be better to use `!topicsStr.trim().isEmpty()`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java]: Also maybe `!topicsRegexStr.trim().isEmpty()`?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (rhauch) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java]: Is the `<p/>` necessary?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java]: Minor, but would using + operator be more consistent with the other lines that make up this literal?

**Review by rhauch — COMMENTED** [review]
A few minor nits, but overall looks really great. Thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (jklukas) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java]: This piece with the `<p/>` is directly copied from https://github.com/apache/kafka/blob/trunk/connect/transforms/src/main/java/org/apache/kafka/connect/transforms/RegexRouter.java#L30-L32, but I can see how Transforms documentation isn't rendered in the same tabular format. Will remove.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (jklukas) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java]: Added a `RegexValidator` here so an exception will be thrown if an invalid regex is passed in. This is used for the analogous config option on `RegexRouter`.

**Comment by jklukas** [comment]
Thanks for the review, @rhauch. All your concerns should be addressed in the two latest commits.

File: docs/connect.html
Comment (rhauch) [inline docs/connect.html]: Minor: the configuration property is just a Java regular expression that _will_ be compiled into a `java.util.regex.Pattern`. I'm not sure it's necessary to denote the implementation, but instead would be sufficient to just say "A Java regular expression of topics to use as input for this connector".

**Review by rhauch — COMMENTED** [review]
One more very minor change. I'd say this looks good, and once the KIP is approved, we can approve this PR and have a committer merge it.

Very nicely done, @jklukas!

**Comment by asfgit** [comment]
FAILURE
 No test results found.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

**Comment by asfgit** [comment]
SUCCESS 
 8051 tests run, 5 skipped, 0 failed.
--none--

**Comment by jklukas** [comment]
I really appreciate the thoughtful review here. It always leads to learning some interesting things.

@ewencp - If you're happy with the KIP as is, you could be the final +1 on the voting thread: https://www.mail-archive.com/dev@kafka.apache.org/msg82225.html

**Comment by jklukas** [comment]
Reading up more on lazy majority, I realize that a KIP needs 3 binding +1 votes, so will need some more eyes. I will let it incubate for a while longer. I'll be on vacation all next week, so I will see if there's been any movement when I get back.

**Comment by jklukas** [comment]
KIP is now approved and as far as I can tell this should be good to merge.

**Comment by rhauch** [comment]
@jklukas let's update the [KIP-215](https://cwiki.apache.org/confluence/display/KAFKA/KIP-215%3A+Add+topic+regex+support+for+Connect+sinks) and [KIP list](https://cwiki.apache.org/confluence/display/KAFKA/Kafka+Improvement+Proposals) pages, and then once recorded as approved we can ask @ewencp to approve and merge this.

**Comment by jklukas** [comment]
I will look through that all next week when I'm back from vacation.

**Comment by jklukas** [comment]
This KIP is now marked as Adopted on the pages @rhauch mentioned.

@ewencp - Does this look ready to merge to you?

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java
Comment (rhauch) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/SinkConnectorConfig.java]: Lines 45 and 46 should use the static constants in this class for the name of the topics: `TOPICS_REGEX_CONFIG ` rather than `SinkTask.TOPICS_REGEX_CONFIG`, and `TOPICS_CONFIG` rather than `SinkTask. TOPICS_CONFIG`.

**Review by rhauch — COMMENTED** [review]
One more minor change.

**Comment by jklukas** [comment]
Pushed a commit to address @rhauch's comment. Should I squash this PR down to a single commit?

**Comment by asfgit** [comment]
FAILURE
 8068 tests run, 5 skipped, 2 failed.
--none--

**Comment by asfgit** [comment]
FAILURE
 8068 tests run, 5 skipped, 2 failed.
--none--

**Review by rhauch — APPROVED** [review]
LGTM. Excellent work, @jklukas, and thanks for pushing this through. (No need to squash, but you can if you want.)

**Comment by jklukas** [comment]
Rebased on trunk and squashed to a single commit. Thanks again for the review on this that allowed it to get into mergeable shape.

**Comment by asfgit** [comment]
FAILURE
 7960 tests run, 5 skipped, 1 failed.
--none--

**Comment by rhauch** [comment]
@junrao or @ewencp, can you please review and merge this into `trunk`? The corresponding [KIP-215](https://cwiki.apache.org/confluence/display/KAFKA/KIP-215%3A+Add+topic+regex+support+for+Connect+sinks) was recently approved, and this PR correctly reflects what's in the KIP. This is a new feature, so it should not be backported. Thanks!

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:283]: What happens with `PatternSyntaxException`?

**Review by ewencp — COMMENTED** [review]
I notice that this is only validating once we're to the point of creating tasks. We also do connector validation elsewhere via the `AbstractHerder.validateConnectorConfig`. We'd probably want to make sure we validate there as well because that prevents a bad config from getting into the system in the first place. We've missed some validations there in the past and it can cause pain because you can have a functioning connector, try to update configs, and then leave it in a broken state. (Obviously possible anyway if you mess up the config, but we'd like to avoid it as much as possible.)

I'm looking back and realizing I think there might be some validations even missing from that code (e.g. I'm not certain we validated non-empty topic config previously), but would it make sense to refactor that code in WorkerSinkTask to be in SinkConnectorConfig and then reuse it in both locations? In fact, it would probably make sense to use it in WorkerConnector.initialize too.

**Review by ewencp — APPROVED** [review]
Ok, I've filed https://issues.apache.org/jira/browse/KAFKA-6253 to follow up on the validation concerns. I'll merge this as is an leave it as follow up.

The PR's tests seem to have gotten stuck in a weird state. I ran unit and integration tests locally to validate.

**Comment by jklukas** [comment]
Thanks @ewencp; I got sidetracked since last week. These are good suggestions and if I have some bandwidth coming up, I'll try to take on 6253.

File: connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java
Comment (jklukas) [reply connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerSinkTask.java:283]: I believe the RegexValidator should already have failed if this pattern could throw an exception.

**Pull request closed without merging**
